lib/src/Uncertainty/Algorithm/WeightedExperiments/SmolyakExperiment.cxx

```diff
@@ -0,0 +1,348 @@
+{
+  LOGDEBUG(OSS() << "SmolyakExperiment::computeNodesAndWeights()");
+  const UnsignedInteger dimension = collection_.getSize();
+  LOGDEBUG(OSS() << "  dimension = " << dimension);
+  if (!(dimension > 0))
+    throw InvalidArgumentException(HERE) << "Error: expected a positive number of marginal experiments, here it is " << dimension;
+  // Compute multiindex set
+  const IndicesCollection combinationIndicesCollection(computeCombination());
+  LOGDEBUG(OSS() << "  combinationIndicesCollection = " << combinationIndicesCollection);
+  // Create elementary Smolyak quadratures
+  Sample nodes(0, dimension);
+  weights.clear();
+  const UnsignedInteger numberOfUnitaryQuadratures = combinationIndicesCollection.getSize();
+  for (UnsignedInteger i = 0; i < numberOfUnitaryQuadratures; ++ i)
+  {
+    WeightedExperimentCollection collection(collection_);
+    for (UnsignedInteger j = 0; j < dimension; ++ j)
+      collection[j].setSize(combinationIndicesCollection(i, j));
+    TensorProductExperiment elementaryExperiment(collection);
+    Point elementaryWeights(0);
+    const Sample elementaryNodes(elementaryExperiment.generateWithWeights(elementaryWeights));
+    // Compute Smolyak coefficient
+    UnsignedInteger marginalLevelsSum = 0;
+    for (UnsignedInteger j = 0; j < dimension; ++ j)
+      marginalLevelsSum += combinationIndicesCollection(i, j);
+    const UnsignedInteger exponent = level_ + dimension - marginalLevelsSum - 1;
+    const Scalar smolyakSign = (exponent % 2 == 0) ? 1.0 : -1.0;
+    const UnsignedInteger binomial = SpecFunc::BinomialCoefficient(dimension - 1, marginalLevelsSum - level_);
+    const Scalar smolyakFactor = smolyakSign * binomial;
+    nodes.add(elementaryNodes);
+    weights.add(smolyakFactor * elementaryWeights);
+  } // Loop over the marginal levels
+  // Reduce to unique nodes and weights
+  if (ResourceMap::GetAsBool("SmolyakExperiment-MergeQuadrature"))
+    mergeNodesAndWeights(nodes, weights);
+  return nodes;
+}
+
+/* Distribution collection accessor */
+SmolyakExperiment::WeightedExperimentCollection SmolyakExperiment::getExperimentCollection() const
+{
+  return collection_;
+}
+
+/* Distribution collection accessor */
+void SmolyakExperiment::setExperimentCollection(const WeightedExperimentCollection & coll)
+{
+  collection_ = coll;
+}
+
+/** Distribution collection accessor */
+void SmolyakExperiment::setLevel(const UnsignedInteger level)
+{
+  level_ = level;
+}
+
+UnsignedInteger SmolyakExperiment::getLevel() const
+{
+  return level_;
+}
+
+/* Size accessor */
+void SmolyakExperiment::setSize(const UnsignedInteger /*size*/)
+{
+  throw NotDefinedException(HERE) << "in SmolyakExperiment::setSize";
+}
+
+UnsignedInteger SmolyakExperiment::getSize() const
+{
+  // only known at generation time
+  return generate().getSize();
+}
+
+/* Compare two points approximately
+   This is for testing purposes only. */
+Bool SmolyakExperiment::ComparePointsApproximately(const Point & x, const Point & y)
```

Comment:
one argument per line

Reply:
ok if we had lots, if only 2 it can stay that way